contracts/tictactoe.scrypt

```diff
@@ -12,38 +17,46 @@ contract TicTacToe {
     //Represents the position of the chessboard. For example, a chess piece with alice in the first row 
     //and first column is expressed as 010000000000000000
     @state
-    bytes board;
+    Piece[N][N] board;
 
     static const int TURNLEN = 1;
-    static const int BOARDLEN = 9;
-    static const bytes EMPTY = b'00';
-    static const bytes ALICE = b'01';
-    static const bytes BOB = b'02';
+    static const int N = 3;
+    static const Piece EMPTY = "";
+    static const Piece ALICE = "X";
+    static const Piece BOB = "O";
+
+    constructor(PubKey alice,  PubKey bob) {
+        this.alice = alice;
+        this.bob = bob;
+        this.is_alice_turn = true;
+        this.board = [["", "", ""],["", "", ""],["", "", ""]];
+    }
+
+
 
-    public function move(int n, Sig sig, int amount, SigHashPreimage txPreimage) {
+    public function move(Location loc, Sig sig, int amount, SigHashPreimage txPreimage) {
 
         require(Tx.checkPreimage(txPreimage));
-        require(n >= 0 && n < BOARDLEN);
 
         // not filled
-        require(Util.getElemAt(this.board, n) == EMPTY);
+        require(this.getPiece(loc.x, loc.y) == EMPTY);
```

Comment:
This top reason for the whole change is to use array indexing directly for write/read.

`require(this.board[n] == EMPTY);`